### ISSUE TITLE
[BUG] Workspace install hangs on invalid inter-dependent packages

### ISSUE THREAD

**Issue by jsg2021** [npm/cli#1920]
### Current Behavior:
npm install hangs

### Expected Behavior:
Either install without hang, or inform user there is a loop in the graph? 

### Steps To Reproduce:
With a workspace defined have two packages depend on each other. In my case, the version spec points to github repos without a hash/fragment. then run npm install on the workspace. Log level silly reveals the looping.

### Environment:
- OS: Fedora 32
- Node: 14.12.0
- npm: 7.0.0.rc.3

**Comment by jsg2021** [comment]
Similar to #1843 but different trigger

**Comment by jsg2021** [comment]
I managed to get out of this state and I haven't been able to create a minimal reproduction. 

There were a few packages that depended on one another and wouldn't install (npm 6) so to make them install we added a "self" dev dep. `"packageName": "file://."` to the two troubled packages. The version ranges in this workspaces use bare git refs for the private packages. We've since fixed the graph to not have the cycle and removed the self-devDeps and, now it's okay. 

Again I have had trouble reproducing this in a minimal set... so I think the tree has to be fairly large with many links. AND a cycle. 

Anyways, I'm going to close this... I'm not sure its a bug on npm.
